File: src/main/java/com/busted_moments/client/config/providers/sound/SoundWriter.java
package com.busted_moments.client.config.providers.sound;

import com.busted_moments.core.config.Config;
import com.busted_moments.core.config.Writer;
import com.busted_moments.core.toml.Toml;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.sounds.SoundEvent;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Type;

@Config.Writer(SoundEvent.class)
public class SoundWriter extends Writer<SoundEvent, String> {
   @Override
   public @Nullable String write(SoundEvent value, Class<?> type, Type... typeArgs) throws Exception {
      return toString(value, type, typeArgs);
   }

   @Override
   public @Nullable SoundEvent read(@NotNull String value, Class<?> type, Type... typeArgs) throws Exception {
      return fromString(value, type, typeArgs);
   }

   @Override
   public SoundEvent fromString(String string, Class<?> type, Type... typeArgs) throws Exception {
      return SoundEvent.createVariableRangeEvent(new ResourceLocation(string));
   }

   @Override
   public String toString(SoundEvent value, Class<?> type, Type... typeArgs) throws Exception {
      return value.getLocation().toString();
   }
}

File: src/main/kotlin/com/busted_moments/client/framework/sounds/SoundWriter.kt
package com.busted_moments.client.framework.sounds

import net.essentuan.esl.encoding.StringBasedEncoder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceLocation
import net.minecraft.sounds.SoundEvent
import java.lang.reflect.AnnotatedElement
import java.lang.reflect.Type
import kotlin.jvm.optionals.getOrNull

object SoundWriter : StringBasedEncoder<SoundEvent>() {
    override fun decode(
        obj: String,
        flags: Set<Any>,
        type: Class<*>,
        element: AnnotatedElement,
        vararg typeArgs: Type
    ): SoundEvent? {
        return BuiltInRegistries.SOUND_EVENT.get(
            ResourceLocation.parse(obj)
        ).getOrNull()?.value()
    }

    override fun encode(
        obj: SoundEvent,
        flags: Set<Any>,
        type: Class<*>,
        element: AnnotatedElement,
        vararg typeArgs: Type
    ): String =
        obj.location.toString()
}